File: Squash.WebAPP/src/app/core/services/url.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { BaseService } from './base/base.service';
import { Url } from '../../models/url.model';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root',
})
export class UrlService extends BaseService<Url> {
  private urlApi = "https://squash-7b6x.onrender.com/api/Url"
  private cachedUrls: Url[] = [];

  getUrlsByUserIdAsync(id: number): Observable<Url[]> {
    return this.http.get<Url[]>(`${this.urlApi}/userId/${id}`);
  }

  getUrlByAliasAsync(alias: string, userId: number): Observable<Url>{
    return this.http.get<Url>(`${this.urlApi}/${alias}/${userId}`);
  }

  urlAliasExistsByIdAsync(alias: string, urlId: number): Observable<boolean>{
    return this.http.get<boolean>(`${this.urlApi}/exist/${alias}/${urlId}`)
  }
  urlAliasExistsAsync(alias: string): Observable<boolean>{
    return this.http.get<boolean>(`${this.urlApi}/exist/${alias}`)
  }

  updateUrlVisitCountAsync(userId:number, urlId: number,  visitCount: number): Observable<boolean>{
    return this.http.put<boolean>(`${this.urlApi}/${userId}/${urlId}/${visitCount}`, {})
  }

  constructor(http: HttpClient) {
    super(http, 'https://squash-7b6x.onrender.com/api/Url');
  }
}

File: Squash.WebAPI/Services/UrlService.cs
using Squash.WebAPI.Interfaces.Repositories;
using Squash.WebAPI.Interfaces.Services;
using Squash.WebAPI.Models;

namespace Squash.WebAPI.Services
{
    public class UrlService(IUrlRepository repository) : IUrlService
    {
        private readonly IUrlRepository _repository = repository;
        public async Task<IEnumerable<Url>> GetAllAsync()
        {
            return await _repository.GetAllAsync();
        }
        public async Task<Url> GetByIdAsync(int id)
        {
            return await _repository.GetByIdAsync(id);
        }
        public async Task<bool> CreateAsync(Url url)
        {
            url.Guid = Guid.NewGuid();
            return await _repository.CreateAsync(url);
        }
        public async Task<bool> DeleteAsync(int id)
        {
            return await _repository.DeleteAsync(id);
        }
        public async Task<bool> UpdateAsync(Url url)
        {
            return await _repository.UpdateAsync(url);
        }
        public async Task<bool> UrlAliasExistsAsync(string alias)
        {
            return await _repository.UrlAliasExistsAsync(alias);
        }
        public async Task<bool> UrlAliasExistsByIdAsync(string alias, int urlId)
        {
            return await _repository.UrlAliasExistsByIdAsync(alias, urlId);
        }
        public async Task<Url> GetUrlByAliasAync(string alias, int userId)
        {
            return await _repository.GetUrlByAliasAync(alias, userId);
        }

        public async Task<IEnumerable<Url>> GetUrlsByUserIdAsync(int userId)
        {
            return await _repository.GetUrlsByUserIdAsync(userId);
        }

        public async Task<bool> UpdateUrlVisitCountAsync(int userId, int urlId, int visitCount) 
        {
            return await _repository.UpdateUrlVisitCountAsync(userId, urlId, visitCount);
        }
    }
}
